src/chatbot/run_chatbot.py

```diff
@@ -45,17 +57,49 @@ def modify_response(response: str, history_len: int) -> str:
     response = response.partition('\n')[0]
     # Remove everything after the last '.', '!', or '?'.
     last_sentence_end_index = max(response.rfind('.'), response.rfind('?'), response.rfind('!'))
-    if last_sentence_end_index != -1:   # -1 means none were found
-        response = response[:last_sentence_end_index+1]
+    if last_sentence_end_index != -1:  # -1 means none were found
+        response = response[:last_sentence_end_index + 1]
     return response.strip()
 
 
-def get_best_response(responses_list: list[str], history_len: int) -> str:
-    # TODO
-    return modify_response(responses_list[0], history_len)
+def get_best_response(responses_list: list[str], history_len: int, prompt_embedding: np.array) -> str:
+    responses = list(map(lambda x: modify_response(x, history_len), responses_list))
+    scored_responses = score_responses(prompt_embedding, responses)
+    if DEBUG:
+        print("----DEBUG-----")
+        for response, score in scored_responses:
+            print(f'{response}: {score}')
+        print("---------")
+    return scored_responses[0][0]
 
 
-model = PapuGaPT2()
+def get_message_tokens(message: str) -> set[str]:
+    tokens = re.findall(r"[\w'\"]+|[,.!?]", message.lower())
+    tokens = [token for token in tokens if len(token) > 2]
+    return set(tokens)
+
+
+def get_embedding(message: str) -> np.array:
+    message_tokens = get_message_tokens(message)
+    tokens_embeddings = []
+    for token in message_tokens:
+        try:
+            token_embedding = embeddings.get_vector(token)
+        except KeyError:
+            continue
+        tokens_embeddings.append(token_embedding)
+    return np.array(tokens_embeddings).mean(axis=0)
+
+
+def score_responses(prompt_embedding: np.array, responses: list[str]) -> list[(str, float)]:
```

Comment:
This is very minor, but if we don't use prompt embedding in main, can't we also get it in this function, and just pass prompt as argument?
It would unify the format between prompt and responses.